hfile.c

```diff
@@ -573,6 +573,9 @@ static hFILE *hopen_fd(const char *filename, const char *mode)
 
 hFILE *hdopen(int fd, const char *mode)
 {
+#if defined HAVE_SETMODE && defined O_BINARY
+    if (setmode(fd, O_BINARY) < 0) return NULL;
+#endif
```

Comment:
[TL;DR: what Rob said.]

As noted in the [#283 commentary](https://github.com/samtools/htslib/pull/283#issuecomment-145835688) that I already pointed you both to, there are at least two problematic cases: `hdopen()` can be too late to set binary mode, and it is invalid to call `setmode()` on some types of file descriptor given to `hdopen()`.

If you are going to call `hdopen()` or `bgzf_dopen()` then it is your responsibility to have opened the file descriptor in binary mode.

The bug is in _test_bgzf.c_'s `try_bgzf_dopen()` which should be setting `O_BINARY` as `hfile_oflags()` does, and probably (being part of htslib so kosher to include _hfile_internal.h_) it should just call `hfile_oflags()` instead of duplicating that code.